runtime/compiler/z/codegen/S390PrivateLinkage.cpp

```diff
@@ -2631,29 +2740,36 @@ J9::Z::PrivateLinkage::buildDirectCall(TR::Node * callNode, TR::SymbolReference
       }
    else
       {
-      if (cg()->getSupportsRuntimeInstrumentation())
-         TR::TreeEvaluator::generateRuntimeInstrumentationOnOffSequence(cg(), TR::InstOpCode::RIOFF, callNode);
-
-      // call through snippet if the method is not resolved or not jitted yet
-      TR::LabelSymbol * label = generateLabelSymbol(cg());
-      TR::Snippet * snippet;
-
-      if (callSymRef->isUnresolved() || (comp()->compileRelocatableCode() && !comp()->getOption(TR_UseSymbolValidationManager)))
+      if (comp()->getGenerateReadOnlyCode())
          {
-         snippet = new (trHeapMemory()) TR::S390UnresolvedCallSnippet(cg(), callNode, label, argSize);
+         gcPoint = buildNoPatchingStaticOrSpecialCall(callNode, callSymRef, dependencies, argSize);
          }
       else
-         {
-         snippet = new (trHeapMemory()) TR::S390J9CallSnippet(cg(), callNode, label, callSymRef, argSize);
-         }
+         { 
+         if (cg()->getSupportsRuntimeInstrumentation())
+            TR::TreeEvaluator::generateRuntimeInstrumentationOnOffSequence(cg(), TR::InstOpCode::RIOFF, callNode);
 
-      cg()->addSnippet(snippet);
+         // call through snippet if the method is not resolved or not jitted yet
+         TR::LabelSymbol * label = generateLabelSymbol(cg());
+         TR::Snippet * snippet;
+
+         if (callSymRef->isUnresolved() || (comp()->compileRelocatableCode() && !comp()->getOption(TR_UseSymbolValidationManager)))
+            {
+            snippet = new (trHeapMemory()) TR::S390UnresolvedCallSnippet(cg(), callNode, label, argSize);
+            }
+         else
+            {
+            snippet = new (trHeapMemory()) TR::S390J9CallSnippet(cg(), callNode, label, callSymRef, argSize);
+            }
 
+         cg()->addSnippet(snippet);
 
-      gcPoint = generateSnippetCall(cg(), callNode, snippet, dependencies, callSymRef);
 
-      if (cg()->getSupportsRuntimeInstrumentation())
-         TR::TreeEvaluator::generateRuntimeInstrumentationOnOffSequence(cg(), TR::InstOpCode::RION, callNode);
+         gcPoint = generateSnippetCall(cg(), callNode, snippet, dependencies, callSymRef);
+
+         if (cg()->getSupportsRuntimeInstrumentation())
+            TR::TreeEvaluator::generateRuntimeInstrumentationOnOffSequence(cg(), TR::InstOpCode::RION, callNode);
```

Comment:
Moved the RION generation outside of the `if/else` , reason I kept that limited to RWX because, I am not sure how RI is going to work out for RX code.